extensions/amp-story/1.0/amp-story.js

```diff
@@ -669,6 +675,23 @@ export class AmpStory extends AMP.BaseElement {
       ) {
         this.systemLayer_.updateProgress(pageId, progress);
       }
+
+      // We can reliably dispatch `PREVIEW_FINISHED` from the `PAGE_PROGRESS`
+      // event's callback because story pages autoplay in preview mode.
+      const lastPageId = this.pages_[this.pages_.length - 1]?.element.id;
+      const isLastPage = pageId === lastPageId;
+      const isPreviewFinished =
+        this.getAmpDoc().isPreview() && isLastPage && progress === 1.0;
+      if (isPreviewFinished && !this.hasDispatchedPreviewFinished_) {
+        dispatch(
```

Comment:
This will dispatch a DOM event, that viewers/players can't access.
Instead, you can use the viewer messenger, here's an example:

https://github.com/ampproject/amphtml/blob/2fb4e151488ba4d856b33f836151d66b45ca457e/extensions/amp-story/1.0/amp-story.js#L1170-L1171